### DISCUSSION TITLE
Ajuster le jour pour éviter les dépassements lors d'une validation avec plusieurs mois

### DISCUSSION THREAD

**Discussion by ageliinas** [MTESSDev/FRW#135] [Annonces]
Bonjour,

Nous avons eu une demande particulière où nous voulons que la date sélectionnée par l'utilisateur soit supérieur à 4 mois à partir de la date d'aujourd'hui. Par exemple; nous sommes le 25 août 2023, le 25 décembre 2023 la date devrait être bonne, car ca fait 4 mois pile. 

Dans le code actuel, c'est à partir du 26 décembre 2023 que la date est bonne. C'est probablement dû au fait que le nombre de jours/semaines est variable d'un mois à l'autre. Existe-t-il une façon d'y remédier?

**Voici la partie de validation du code actuel (celui que la date est bonne à partir du 26 décembre):**
```
validations:
                maxDeuxAnsCustom: 1
                minQuatreMoisCustom: 1
              additionals:
                :validation-rules: |
                  {
                    maxDeuxAnsCustom: ({ value }) => {
                      const currentDate = new Date();
                      const selectedDate = new Date(value);

                      // Ajouter deux ans à la date actuelle
                      currentDate.setFullYear(currentDate.getFullYear() + 2);

                      return selectedDate <= currentDate;
                    },
                    minQuatreMoisCustom: ({ value }) => {
                      const currentDate = new Date();
                      const selectedDate = new Date(value);

                      // Ajouter quatre mois à la date actuelle
                      currentDate.setMonth(currentDate.getMonth() + 4);

                      return selectedDate >= currentDate;
                    }
                  }
              validation-messages:
                maxDeuxAnsCustom:
                  fr: |
                    ({ name }) => {
                      return `La valeur du champ « ${name} » ne doit pas dépasser deux ans à partir de la date d'aujourd'hui.`;
                    }
                minQuatreMoisCustom:
                  fr: |
                    ({ name }) => {
                      return `La valeur du champ « ${name} » doit être au minimum quatre mois après la date d'aujourd'hui.`;
                    }
``` 

**Nous avions aussi essayé avec un computed pour ajuster le jour, mais ca n'a pas fonctionné;
Code du computed:**

```
monthAdd:
        code: |
          (){ 
              const currentDate = new Date(); // Date actuelle
              const monthsToAdd = 4; // Nombre de mois à ajouter

              let temp = new Date(currentDate.getFullYear(), currentDate.getMonth(), 1);
              temp.setMonth(temp.getMonth() + (monthsToAdd + 1));
              temp.setDate(temp.getDate() - 1); 

              if (currentDate.getDate() < temp.getDate()) { 
                temp.setDate(currentDate.getDate()); 
              }

              return temp;
            }
``` 

**Code de la validation en lien avec le computed:**

```
validations:
                maxDeuxAnsCustom: 1
                minQuatreMoisCustom: 1
              additionals:
                :validation-rules: |
                  {
                    maxDeuxAnsCustom: ({ value }) => {
                      const currentDate = new Date();
                      const selectedDate = new Date(value);

                      // Ajouter deux ans à la date actuelle
                      currentDate.setFullYear(currentDate.getFullYear() + 2);

                      return selectedDate <= currentDate;
                    },
                    minQuatreMoisCustom: ({ value }) => {
                      const currentDate = new Date(); // Date actuelle
                      const selectedDate = new Date(value); // Date sélectionnée dans le champ

                      const monthsToAdd = 4; // Nombre de mois à ajouter

                      const minDate = monthAdd(currentDate, monthsToAdd);

                      return selectedDate >= minDate; // Vérifier si la date sélectionnée est supérieure ou égale à la date calculée
                    }                    
                  }
              validation-messages:
                maxDeuxAnsCustom:
                  fr: |
                    ({ name }) => {
                      return `La valeur du champ « ${name} » ne doit pas dépasser deux ans à partir de la date d'aujourd'hui.`;
                    }
                minQuatreMoisCustom:
                  fr: |
                    ({ name }) => {
                      return `La valeur du champ « ${name} » doit être au minimum quatre mois après la date d'aujourd'hui.`;
                    }
``` 

Merci! :)

**Comment by j0ber** [comment]
Bonjour,

Je crois que c'est simplement votre utilisation du constructeur Date qui n'est pas bonne.

Lorsqu'on utilise une string en paramètre, il faut avoir l'heure. Sinon la date obtenue correspondra à la journée précédente. Exemple '2023-12-25' sans heure va donner la date '2023-12-24'.

Il faudrait donc initialiser votre date comme ceci : 
` const selectedDate = new Date(value + 'T00:00:00')`

Référence: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date

N'oubliez pas que vous pouvez mettre des console.log  et debugger dans votre code afin de "debugger" dans le fureteur une fois le formulaire déployé.

**Reply by ageliinas** [reply]
Bonjour,

J'ai changé le const selectedDate = new Date(value); pour const selectedDate = new Date(value + 'T00:00:00') comme mentionné, mais le problème persiste toujours; par rapport à la date d'aujourd'hui, ce serait à partir du 28 décembre que le tout est bon (4 mois), mais ce n'est toujours pas le cas. Il considère encore que la bonne date est à partir du 29 décembre dans ce cas-ci, soit la journée suivante.

**Comment by j0ber** [comment]
en fait, il faut mettre la fin de la journée, car vous comparez des dates complètes et que selectedDate va contenir  minuit comme heure, ce qui sera toujours plus petit que la date que vous calculez à partir de la date du jour qui contient l'heure courante.

Si vous voulez travailler avec des dates complètes il faudrait ceci : 
`selectedDate = new Date(value + 'T23:59:59.999Z')`

Sinon il faut travailler avec une string qui contient uniquement AAAA-MM-JJ, du style : `maDate.toISOString().slice(0, 10)`

**Reply by ageliinas** [reply]
Merci ca fonctionne maintenant!
